File: ZigZag Conversion/ZigZag Conversion.java
public class Solution {
    public String convert(String s, int nRows) {
        if (nRows <= 1) return s;
        final int key = 2 * nRows - 2;
        final int length = s.length();
        char[] c = s.toCharArray();
        StringBuilder res= new StringBuilder("");
        for (int i = 0; i < nRows; ++i) {
            for (int j = i; j < length; j += key) {
                res.append(c[j]);
                if (i > 0 && i < nRows - 1 && j + key - 2 * i < length) res.append(c[j + key - 2 * i]); 
            }
        }
        return res.toString();
    }
}

File: ZigZag Conversion/ZigZag Conversion.cpp
class Solution {
public:
    string convert(string s, int nRows) {
        if (nRows <= 1) return s;
        const int key = 2 * nRows - 2;
        const int size = s.size();
        string res("");
        for (int i = 0; i < nRows; ++i) {
            for (int j = i; j < size; j += key) {
                res += s[j];
                if (i > 0 && i < nRows - 1 && j + key - 2 * i < size) res += s[j + key - 2 * i]; 
            }
        }
        return res;
    }
};